plumbing/format/packfile/encoder.go

```diff
@@ -0,0 +1,115 @@
+package packfile
+
+import (
+	"crypto/sha1"
+	"fmt"
+	"hash"
+	"io"
+
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/plumbing/storer"
+	"gopkg.in/src-d/go-git.v4/utils/binary"
+
+	"github.com/klauspost/compress/zlib"
+)
+
+// Encoder gets the data from the storage and write it into the writer in PACK
+// format
+type Encoder struct {
+	storage storer.ObjectStorer
+	w       io.Writer
+	hash    hash.Hash
+}
+
+// NewEncoder creates a new packfile encoder using a specific Writer and
+// ObjectStorer
+func NewEncoder(w io.Writer, s storer.ObjectStorer) *Encoder {
+	h := sha1.New()
+	mw := io.MultiWriter(w, h)
+	return &Encoder{
+		storage: s,
+		w:       mw,
+		hash:    h,
+	}
+}
+
+// Encode encodes objects specified using a list of hashes. This objects must
+// exists into the storer
+func (e *Encoder) Encode(hashes []plumbing.Hash) (plumbing.Hash, error) {
```

Comment:
From the signature of the method and the comments it is not clear if you can create a **single** packfile by calling several times in a row to **Encode**.

I would change the comment here to clarify that every time you call Encode it **creates a new packfile** and write it to the writer. Something like this:

```
// Encode creates a packfile containing all the objects referenced in hashes and writes it to the writer in the Encoder.
```